### ISSUE TITLE
[New Edit Mode Prototype] Switching view mode triggers MCTBeforeUnload confirmation.

### ISSUE THREAD

**Issue by akhenry** [nasa/openmct#351]
This has the effect of showing the user the onunload dialog. Might be due to `view=...` being appended to the url.

**Comment by akhenry** [comment]
For now, have just disabled the URL rewriting on view change in `BrowseObjectController.js` when in edit mode. This actually makes it consistent with the behavior in old edit mode. Having tested it, it seems to work without side-effects. 

@VWoeltjen what is the purpose of the URL rewriting on view change, is it purely cosmetic ( to keep the URL in sync with the view state), or is there other functionality somewhere that is dependent on this URL rewrite?

**Comment by VWoeltjen** [comment]
I may be misunderstanding the question.

The view gets [set in the URL](https://github.com/nasa/openmctweb/blob/d6fe543c16ba32c637b2106aaf3da07e0e49929b/platform/commonUI/browse/src/BrowseObjectController.js#L56) such that view selection can be bookmarked, linked-to, etc.

There is also a [funny workaround](https://github.com/nasa/openmctweb/blob/d6fe543c16ba32c637b2106aaf3da07e0e49929b/platform/commonUI/browse/src/BrowseObjectController.js#L57-L64) to default Angular behavior. Setting the view in the location hash is a location change, prompting `ngRoute`/`ng-view` to do their business, which would normally have the side effect of re-instantiating all controllers from the top level on down. This means that any application state that _isn't_ captured in the URL (e.g. tree expand/collapse state) would be lost, which isn't what we want. The single-use event listener there "tricks" Angular into thinking there isn't a real route change - in this case, we are relying on the fact that the application is in the correct state already with respect to the URL.

**Comment by larkin** [comment]
Disabling URL rewrite in edit mode makes perfect sense for the prototype.

Long term, I think we remove view switching during edit mode.  Only the telemetry panel makes use of it, and we've discussed phasing it out, as it confuses users and results in overly complicated view objects.

**Comment by akhenry** [comment]
If you misunderstood the question, it's because I didn't really explain it very well :) 

Yep, understand why the view is encoded in the URL. Ok, so the code you link to is intended to actively prevent controllers etc. from re-initializing? Would reloadOnSearch=false not have worked for this? Or is it because you want it to reload for everything _except_ view change?

To try and explain my original question in a clearer way, the act of setting the "view" query parameter in the URL causes $locationChangeSuccess to be fired, which triggers a confirmation popup. The logic that does this is in MCTBeforeUnload, so I don't want to put code in there that is specific to view state change. Instead I've just added a line in `BrowseObjectController` that checks if we're in edit mode, and if so does not set the view state on the URL when it changes. This was not a problem in old edit mode, because it the edit mode controller was not actively keeping the URL in sync with the view state (either through oversight or by design for for this very reason, I'm not sure).

The change does not seems to have had any ill effect, but I wanted to check whether there are any potential knock on effects that I haven't considered. There is in fact one that I've discovered - if you begin editing in plot mode, then switch to fixed layout, then hit save, you will be taken back to plot mode. This is as a direct consequence of the URL not being rewritten in edit mode.

@larkin But we are keeping fixed layouts, just moving them from panels to layouts, no?

**Comment by akhenry** [comment]
* updated the title to more accurately reflect the issue

**Comment by VWoeltjen** [comment]
Agreed with @larkin that whatever workaround works is fine for prototyping.

> Would reloadOnSearch=false not have worked for this? Or is it because you want it to reload for everything except view change?

Poking around the commit history I see that `reloadOnSearch` is included (and was included at that time.) Not sure why the extra behavior on `$locationChangeSuccess` was added there; my recollection here is that `reloadOnSearch` did not behave as expected in Angular 1.2.27, but we've since upgraded to 1.4.4. If I comment out that block now, I can change the view without the tree collapsing.

> To try and explain my original question in a clearer way, the act of setting the "view" query parameter in the URL causes $locationChangeSuccess to be fired, which triggers a confirmation popup. The logic that does this is in MCTBeforeUnload, so I don't want to put code in there that is specific to view state change. Instead I've just added a line in BrowseObjectController that checks if we're in edit mode, and if so does not set the view state on the URL when it changes. This was not a problem in old edit mode, because it the edit mode controller was not actively keeping the URL in sync with the view state (either through oversight or by design for for this very reason, I'm not sure).

I'm guessing that removing the `$locationChangeSuccess` listener from BrowseObjectController won't change the problem here.

I think it makes sense that `mct-before-unload` should fire on route changes for consistency; don't want special case behavior in there, don't want logic that will fail to catch warning-worthy changes in the future.

One option would be to add an extra attribute to `mct-before-unload` that evaluates route changes to see if they're warning-worthy or not. To keep the interface small, `mct-before-unload` could also pass some event object into the expression, and Edit mode could inspect this to see if it needs to issue a warning. (Is it a route change? Yes. Is the only thing that changed the view? Yes. Return undefined - don't need to prevent that.)
